[JIT] Add mutation checks for tensor inputs

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79078

- Added mutation checks in SchemaCheckTensor that raise RuntimeError if an argument is mutated when mutation is not specified by the matching schema
- Added python binding for an is_mutable variable in python Argument class
- Tested for correct functionality in operator chains where:
    -  a default argument is replaced
    - a mutable operator is used
    - an operator mutates non-specified mutable argument  
    - an operator chain mutates a non specified mutable argument

**Note that functionality in operators which have Tensor[] parameters and operators which have tensors as key word arguments are not tested in this pr, but will be covered in a later pr.**

Commit:
Add mutation checks for tensor inputs

[ghstack-poisoned]

Krovatkin commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -38,16 +43,69 @@ def __repr__(self):
-            return e.elem if isinstance(e, cls) else e
+            if isinstance(e, cls):
+                return e.elem
+            elif is_iterable_of_dtype(e, cls):
```
Isn't `tree_map` supposed to take care of this case and recursively traverse lists and dicts?

Krovatkin commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -38,16 +43,69 @@ def __repr__(self):
+
+        for argument in schema.arguments:
+            name = argument.name if argument.name != "self" else "input"
+            if (arguments.get(name) is not None and cloned_arguments.get(name) is not None):
```
isn't it enough to check `arguments.get(name)`?
`cloned_arguments.get(name)` should be an assert?

Krovatkin commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1243,29 +1243,34 @@ def is_iterable(obj):
 
     return True
 
+def is_iterable_of_tensors(iterable, include_empty=False):
```
where is this used?

Krovatkin commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -38,16 +43,69 @@ def __repr__(self):
+
+        if (len(matched_schemas) == 0):
+            raise RuntimeError('No matching schema found')
+            return None
```
`return None` is dead code?

Krovatkin commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -38,16 +43,69 @@ def __repr__(self):
+            return matched_schemas[0]
+        else:
+            raise RuntimeError('More than one matching schema found')
+            return None
```
ditto

Krovatkin commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -1,11 +1,16 @@
+from torch.testing._internal.common_utils import is_iterable_of_dtype
+from typing import Any, Callable, Dict, Tuple
 
 schema_check_recorded_ops = []
```
I feel a bit wary of globals. Can't we just store the sequence of ops on a tensor subclass object?
There's some potential for data races here.

Krovatkin commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -38,16 +43,69 @@ def __repr__(self):
+                if (len(before_list) != len(after_list)):
+                    return True
+                for before_elem, after_elem in zip(before_list, after_list):
+                    if (not torch.equal(before_elem, after_elem)):
```
do we need to handle dicts for any ops?

goldenxuett replied:
If you mean dictionaries with tensors as values as arguments, I'm not sure if any ops have a signature like that, I can either look into that or just implement a check for that case.

Krovatkin commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -38,16 +43,69 @@ def __repr__(self):
+                continue
+
+        if (len(matched_schemas) == 0):
+            raise RuntimeError('No matching schema found')
```
this is a great check!

eellison commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -38,16 +43,69 @@ def __repr__(self):
+        def has_mutated(before, after):
+            if isinstance(before, torch.Tensor) and isinstance(after, torch.Tensor):
+                return not torch.equal(before, after)
+            elif is_iterable_of_dtype(before, torch.Tensor) and is_iterable_of_dtype(after, torch.Tensor):
```
I would look into `tree_map` and `tree_iterate` to handle nested containers

Commit:
Update on "[JIT] Add mutation checks for tensor inputs"


- Added mutation checks in SchemaCheckTensor that raise RuntimeError if an argument is mutated when mutation is not specified by the matching schema
- Created more general is_iterable_of_dtype() in common_utils
- Added python binding for an is_mutable variable in python Argument class
- Tested for correct functionality in operator chains where:
    -  a default argument is replaced
    - a mutable operator is used
    - an operator mutates non-specified mutable argument  
    - an operator chain mutates a non specified mutable argument

**Note that functionality in operators which have Tensor[] parameters and operators which have tensors as key word arguments are not tested in this pr, but will be covered in a later pr.**

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -43,14 +55,38 @@ def unwrap(e):
-def reset_cache():
-    global schema_check_recorded_ops
-    schema_check_recorded_ops.clear()
+def find_matching_schema(op: Callable, args : Tuple[Any], kwargs : Dict[str, Any]):
```
Maybe @eellison would know better, but I think you can actually skip all the indirection here around `get_signature_for_torch_op`.

The idea with `get_signature_for_torch_op` is that if you have a torch operator like `torch.add`, it actually represents a bundle of operator overloads that each have separate schemas (for add, it would be `add.Tensor`, `add.Scalar` and `add.out`).

If you only had the bundle, then you'd be forced to do these lookups. But you're inside of a `__torch_dispatch__` call, so you're actually in a better situation: the `func` object passed into `__torch_dispatch__` represents a single overload. For example, if you called `torch.add`, then the `func` by the time you hit `__torch_dispatch__` would be `torch.ops.aten.add.Tensor`.

Since those overload objects represent individual overloads, then they have a unique schema. You can access it directly with `torch.ops.aten.add.Tensor._schema`, which returns a `FunctionSchema` object. I think that all the information that you need should be on there, but if not, we should probably just expose it to python.

goldenxuett replied:
@bdhirsh I think `._schema` is missing the information about which arguments map to which values specifically due to the possibility of having both kwargs and args out of order. Is there a way to find this mapping from the `FunctionSchema` object without the signature? Because part of how I was using `get_signature_for_torch_op` was to get this mapping using `bind`

bdhirsh commented on torch/testing/_internal/schema_check_tensor.py:
```diff
@@ -30,6 +34,14 @@ def __new__(cls, elem):
 
+    @staticmethod
+    def reset_cache():
+        SchemaCheckTensor.recorded_ops.clear()
```
Any reason not to make the recorded ops per-instance? It seems maybe more intuitive that if you have multiple `SchemaCheckTensors` floating around, that they would only record the ops that they were run on (instead of sharing / modifying a shared list).

bdhirsh replied:
(just realized I'm basically echo'ing what Nick said :) )

eellison reviewed (commented):
(not needed in this pr) would be nice to refactor this to also cover Tensor Metadata changing (storage, strides, size) like https://github.com/pytorch/pytorch/pull/77313/files#diff-d183f2afc51d6a59bc70094e8f476d2468c45e415500f6eb60abad955e065156R1419 is doing.

Right now something like `resize_` isn't being tested because the output and input will be equal

Commit:
Update on "[JIT] Add mutation checks for tensor inputs"


- Added mutation checks in SchemaCheckTensor that raise RuntimeError if an argument is mutated when mutation is not specified by the matching schema
- Added python binding for an is_mutable variable in python Argument class
- Tested for correct functionality in operator chains where:
    -  a default argument is replaced
    - a mutable operator is used
    - an operator mutates non-specified mutable argument  
    - an operator chain mutates a non specified mutable argument

**Note that functionality in operators which have Tensor[] parameters and operators which have tensors as key word arguments are not tested in this pr, but will be covered in a later pr.**

[ghstack-poisoned]

davidberard98 commented on test/jit/test_schema_check.py:
```diff
@@ -18,21 +18,51 @@
+
+    # Tests that an exception is raised for a mismatching mutation
+    def test_mutation_check_fail(self):
+        with self.assertRaises(RuntimeError):
```
nice test!!

davidberard98 reviewed (approved):
lgtm!

Krovatkin reviewed (approved):
:shipit:

goldenxuett commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot revert -m "broke bazel build-and-test"

davidberard98 commented:
@pytorchbot revert -m "broke bazel build-and-test, see https://github.com/pytorch/pytorch/runs/6836001002?check_suite_focus=true"-c ignoredsignal

davidberard98 commented:
@pytorchbot revert -m "broke bazel build-and-test, see [https://github.com/pytorch/pytorch/runs/6836001002?check_suite_focus=true](https://github.com/pytorch/pytorch/runs/6836001002?check_suite_focus=true%22)" -c ignoredsignal
